Introduce PublicSuffixStore with cache

szewai opened the pull request:
#### ac05e4cafed5aa893fdfb19a86d27ad1583ef366
<pre>
Introduce PublicSuffixStore with cache
<a href="https://bugs.webkit.org/show_bug.cgi?id=271188">https://bugs.webkit.org/show_bug.cgi?id=271188</a>
<a href="https://rdar.apple.com/124971859">rdar://124971859</a>

Reviewed by Per Arne Vollan and Chris Dumez.

Intoduce a new class PublicSuffixStore and move existing public suffix related functions into it. PublicSuffixStore can
have a cache for public suffix. When the cache is in use, PublicSuffixStore searches public suffix in it before invoking
platform specific code (this may help us to tighten sandbox in some process later). Currently the cache is only enabled
in web process on Cocoa platforms.

Before this patch, web process has a cache for finding top privately controlled domain of a given host. UI process sends
(host, domain) pair to web process on load time, and web process will add the entry in cache (256605@main). This patch
makes UI process send the public suffix the to web process, and makes web process store it in the new public suffix
cache, since the domain cache has a size limit, and it only benefits search for the given host. Also, this patch fixes
the regression from 270762@main, that web process does not update the cache on receiving LoadParameters (since
setTopPrivatelyControlledDomain is dropped from LoadParameters::decode), by adding addPublicSuffix to
WebPage::platformDidReceiveLoadParameters.

* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/SourcesCocoa.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::urlForBindings const):
* Source/WebCore/loader/ResourceLoadStatistics.cpp:
* Source/WebCore/loader/cache/MemoryCache.cpp:
* Source/WebCore/page/Quirks.cpp:
(WebCore::isYahooMail):
(WebCore::Quirks::shouldHideSearchFieldResultsButton const):
(WebCore::Quirks::isAmazon const):
(WebCore::Quirks::isGoogleMaps const):
* Source/WebCore/page/SecurityOrigin.cpp:
(WebCore::SecurityOrigin::isSameSiteAs const):
(WebCore::SecurityOrigin::isMatchingRegistrableDomainSuffix const):
* Source/WebCore/page/csp/ContentSecurityPolicySourceList.cpp:
(WebCore::ContentSecurityPolicySourceList::isValidSourceForExtensionMode):
* Source/WebCore/platform/Curl.cmake:
* Source/WebCore/platform/PublicSuffixStore.cpp: Added.
(WebCore::PublicSuffixStore::singleton):
(WebCore::PublicSuffixStore::clearHostTopPrivatelyControlledDomainCache):
(WebCore::PublicSuffixStore::isPublicSuffix const):
(WebCore::PublicSuffixStore::publicSuffix const):
(WebCore::PublicSuffixStore::topPrivatelyControlledDomain const):
* Source/WebCore/platform/PublicSuffixStore.h: Renamed from Source/WebCore/platform/PublicSuffix.h.
* Source/WebCore/platform/RegistrableDomain.h:
(WebCore::RegistrableDomain::uncheckedCreateFromHost):
(WebCore::RegistrableDomain::registrableDomainFromHost):
* Source/WebCore/platform/SourcesSoup.txt:
* Source/WebCore/platform/cocoa/PublicSuffixCocoa.mm: Removed.
* Source/WebCore/platform/cocoa/PublicSuffixStoreCocoa.mm: Added.
(WebCore::isPublicSuffixCF):
(WebCore::PublicSuffixStore::platformIsPublicSuffix const):
(WebCore::PublicSuffixStore::platformTopPrivatelyControlledDomain const):
(WebCore::PublicSuffixStore::enablePublicSuffixCache):
(WebCore::PublicSuffixStore::addPublicSuffix):
* Source/WebCore/platform/glib/UserAgentQuirks.cpp:
(WebCore::UserAgentQuirks::quirksForURL):
* Source/WebCore/platform/graphics/skia/PathSkia.cpp:
* Source/WebCore/platform/network/NetworkStorageSession.cpp:
* Source/WebCore/platform/network/ResourceRequestBase.cpp:
(WebCore::ResourceRequestBase::partitionName):
* Source/WebCore/platform/network/cf/NetworkStorageSessionCFNet.cpp:
* Source/WebCore/platform/network/cf/ResourceRequestCFNet.cpp:
* Source/WebCore/platform/network/curl/CookieJarDB.cpp:
(WebCore::CookieJarDB::hasCookies):
(WebCore::CookieJarDB::canAcceptCookie):
* Source/WebCore/platform/network/curl/PublicSuffixStoreCurl.cpp: Renamed from Source/WebCore/platform/network/curl/PublicSuffixCurl.cpp.
(WebCore::PublicSuffixStore::platformIsPublicSuffix const):
(WebCore::topPrivatelyControlledDomainInternal):
(WebCore::PublicSuffixStore::platformTopPrivatelyControlledDomain const):
* Source/WebCore/platform/soup/PublicSuffixStoreSoup.cpp: Renamed from Source/WebCore/platform/soup/PublicSuffixSoup.cpp.
(WebCore::PublicSuffixStore::platformIsPublicSuffix const):
(WebCore::permissiveTopPrivateDomain):
(WebCore::PublicSuffixStore::platformTopPrivatelyControlledDomain const):
* Source/WebKit/NetworkProcess/cocoa/NetworkProcessCocoa.mm:
* Source/WebKit/NetworkProcess/soup/NetworkDataTaskSoup.cpp:
(WebKit::NetworkDataTaskSoup::shouldAllowHSTSPolicySetting const):
* Source/WebKit/Shared/GoToBackForwardItemParameters.h:
* Source/WebKit/Shared/GoToBackForwardItemParameters.serialization.in:
* Source/WebKit/Shared/LoadParameters.h:
* Source/WebKit/Shared/LoadParameters.serialization.in:
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionMatchPatternCocoa.mm:
(WebKit::WebExtensionMatchPattern::hostIsPublicSuffix const):
* Source/WebKit/UIProcess/ProvisionalPageProxy.cpp:
(WebKit::ProvisionalPageProxy::goToBackForwardItem):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::launchProcessForReload):
(WebKit::WebPageProxy::loadRequestWithNavigationShared):
(WebKit::WebPageProxy::loadFile):
(WebKit::WebPageProxy::goToBackForwardItem):
(WebKit::WebPageProxy::resetStateAfterProcessTermination):
* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::WebProcessProxy::processDidTerminateOrFailedToLaunch):
(WebKit::WebProcessProxy::didExceedMemoryFootprintThreshold):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataRecord.cpp:
(WebKit::WebsiteDataRecord::displayNameForHostName):
(WebKit::WebsiteDataRecord::displayNameForOrigin):
(WebKit::WebsiteDataRecord::topPrivatelyControlledDomain):
* Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm:
(WebKit::WebPage::platformDidReceiveLoadParameters):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::goToBackForwardItem):
* Source/WebKit/WebProcess/cocoa/WebProcessCocoa.mm:
(WebKit::prewarmLogs):
(WebKit::WebProcess::platformInitializeProcess):
* Tools/TestWebKitAPI/Tests/WebCore/PublicSuffix.cpp:
(TestWebKitAPI::TEST_F):

Canonical link: <a href="https://commits.webkit.org/276505@main">https://commits.webkit.org/276505@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7894d1d11e456d890b136bd8ce215892ae027af9

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/44831 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/23929 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/47320 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/47485 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/40836 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/47134 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/27978 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/21325 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/36820 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/45409 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/20988 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/38606 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/17893 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/18406 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/39741 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/2878 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/41063 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/40038 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/49149 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/19798 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/16353 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/43808 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/21117 "Built successfully") | | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/42564 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/9985 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/21456 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/20792 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

annevk commented on Source/WebCore/platform/PublicSuffix.cpp:
```diff
@@ -0,0 +1,116 @@
+
+String PublicSuffixStore::publicSuffix(const String& host)
+{
+    if (host.isEmpty() || !host.containsOnlyASCII())
```
We should really move towards being able to assert this as the URL parser will ensure a host is pure ASCII. I guess this preserves the existing contract so maybe add a FIXME?

(Same concern applies below to topPrivatelyControlledDomain.)

szewai replied:
Yes this is the existing check. I think the problem is with existing interface, there is no guarantee that the host parameter is result from URL parser (and we use plain string in tests so if I drop it the test fails.). 
It might be better to use URL as parameter so we could avoid these checks. I will add a FIXME here.

annevk commented on Source/WebCore/platform/PublicSuffix.cpp:
```diff
@@ -0,0 +1,116 @@
+    if (host.isEmpty() || !host.containsOnlyASCII())
+        return { };
+
+    const auto lowercaseHost = host.convertToASCIILowercase();
```
This too shouldn't be needed. The host should already be in canonical form at this point.

szewai replied:
As explained above, if I drop the check the test fails as it uses crafted string instead of result form URL parser. I will fix the test and the checks in followup.

annevk commented on Source/WebCore/platform/PublicSuffix.cpp:
```diff
@@ -0,0 +1,116 @@
+    return { };
+}
+
+String PublicSuffixStore::topPrivatelyControlledDomain(const String& host)
```
While we're changing how to get to this method, should we take this opportunity to rename it to `registrableDomain()` to match the standard?

https://url.spec.whatwg.org/#host-registrable-domain

szewai replied:
It seems we actually have a `registrableDomainFromHost` in RegistrableDomain.h, and I think we could merge these functions too. 
I will work on the renaming as followup to make this patch easier to review.

annevk commented on Source/WebCore/platform/network/ResourceRequestBase.cpp:
```diff
@@ -829,7 +829,7 @@ String ResourceRequestBase::partitionName(const String& domain)
     if (domain.isNull())
         return emptyString();
-    String highLevel = topPrivatelyControlledDomain(domain);
+    String highLevel = PublicSuffixStore::singleton().topPrivatelyControlledDomain(domain);
```
auto*

annevk commented on Source/WebKit/Shared/GoToBackForwardItemParameters.h:
```diff
@@ -43,7 +43,7 @@ struct GoToBackForwardItemParameters {
     bool lastNavigationWasAppInitiated;
     std::optional<NetworkResourceLoadIdentifier> existingNetworkResourceLoadIdentifierToResume;
-    std::optional<String> topPrivatelyControlledDomain;
+    String publicSuffix;
```
Thanks for cleaning this up! I was wondering if we could make a change like this after my last PR.

annevk commented on Source/WebKit/UIProcess/ProvisionalPageProxy.cpp:
```diff
@@ -286,15 +286,14 @@ void ProvisionalPageProxy::goToBackForwardItem(API::Navigation& navigation, WebB
 
-    auto topPrivatelyControlledDomain = WebCore::topPrivatelyControlledDomain(URL(item.url()).host().toString());
-
+    String publicSuffix = WebCore::PublicSuffixStore::singleton().publicSuffix(URL(item.url()).host().toString());
```
auto*

annevk commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -2232,9 +2230,8 @@ RefPtr<API::Navigation> WebPageProxy::goToBackForwardItem(WebBackForwardListItem
-    auto topPrivatelyControlledDomain = WebCore::topPrivatelyControlledDomain(URL(item.url()).host().toString());
-    
-    send(Messages::WebPage::GoToBackForwardItem({ navigation->navigationID(), item.itemID(), frameLoadType, ShouldTreatAsContinuingLoad::No, std::nullopt, m_lastNavigationWasAppInitiated, std::nullopt, topPrivatelyControlledDomain, { } }));
+    String publicSuffix = WebCore::PublicSuffixStore::singleton().publicSuffix(URL(item.url()).host().toString());
```
auto*

annevk commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -9498,7 +9495,7 @@ void WebPageProxy::resetStateAfterProcessTermination(ProcessTerminationReason re
 #if PLATFORM(IOS_FAMILY)
     if (m_process->isUnderMemoryPressure()) {
-        String domain = WebCore::topPrivatelyControlledDomain(URL({ }, currentURL()).host().toString());
+        String domain = WebCore::PublicSuffixStore::singleton().topPrivatelyControlledDomain(URL({ }, currentURL()).host().toString());
```
auto*

annevk commented on Source/WebKit/UIProcess/WebProcessProxy.cpp:
```diff
@@ -1199,7 +1199,7 @@ void WebProcessProxy::processDidTerminateOrFailedToLaunch(ProcessTerminationReas
     if (pages.size() == 1 && reason == ProcessTerminationReason::Crash) {
         auto& page = pages[0];
-        String domain = topPrivatelyControlledDomain(URL({ }, page->currentURL()).host().toString());
+        String domain = PublicSuffixStore::singleton().topPrivatelyControlledDomain(URL({ }, page->currentURL()).host().toString());
```
auto*

annevk commented on Source/WebKit/UIProcess/WebProcessProxy.cpp:
```diff
@@ -2008,7 +2008,7 @@ void WebProcessProxy::didExceedMemoryFootprintThreshold(size_t footprint)
 
     for (auto& page : this->pages()) {
-        String pageDomain = topPrivatelyControlledDomain(URL({ }, page->currentURL()).host().toString());
+        String pageDomain = PublicSuffixStore::singleton().topPrivatelyControlledDomain(URL({ }, page->currentURL()).host().toString());
```
auto*

weinig commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -4065,8 +4065,9 @@ const URL& Document::urlForBindings() const
         auto areSameSiteIgnoringPublicSuffix = [](StringView domain, StringView otherDomain) {
-            auto domainString = topPrivatelyControlledDomain(domain.toStringWithoutCopying());
-            auto otherDomainString = topPrivatelyControlledDomain(otherDomain.toStringWithoutCopying());
+            auto& publicSuffixStore = PublicSuffixStore::singleton();
```
I would recommend against using a singleton and instead hang an instance off the Page. Singletons, while appealing in their simplicity, make testing and reuse more challenging.

szewai replied:
I am not sure about "hang an instance off the Page"; could you give an example? Like having Ref<PublicSuffixStore> in PageConfiguration? If so, it seems workers will not work?

cdumez replied:
The code was indeed called from various threads. However, the old code was thread safe, the new one is not (see my other comment). Public Suffix was always a global static map with a Lock. So using a singleton doesn't make things worse in that aspect (it was always global). However, I'm curious what you feel that benefits are to using a singleton vs keeping the current global static map?

szewai replied:
The goal of this change is to introduce a public suffix cache but use it only in certain case (web process in cocoa platforms). Having a singleton makes the configuration and management easier (compared to adding a global flag (or a global cache) for controlling the caching behavior).

weinig commented on Source/WebCore/platform/PublicSuffix.h:
```diff
@@ -25,13 +25,33 @@
+    Lock m_HostTopPrivatelyControlledDomainCacheLock;
+    HashMap<String, String, ASCIICaseInsensitiveHash> m_hostTopPrivatelyControlledDomainCache WTF_GUARDED_BY_LOCK(m_HostTopPrivatelyControlledDomainCacheLock);
+    Lock m_publicSuffixCacheLock;
+    std::optional<HashSet<String, ASCIICaseInsensitiveHash>> m_publicSuffixCache WTF_GUARDED_BY_LOCK(m_publicSuffixCacheLock);
```
I am not sure `cache` is the word you are going for here as this isn't something that can be invalidated and still yield correct results.

From use, it seems like this is more of an "additionalPublicSuffixes" set.

szewai replied:
I used cache as it's stored in memory, while the platform functions might look it up in a file. The cache entries is a subset of the file, and they are unlikely to change as the public suffix list is static (maybe we want to evict some entries as memory optimization later).

weinig commented on Source/WebCore/platform/PublicSuffix.h:
```diff
@@ -25,13 +25,33 @@
+    WEBCORE_EXPORT static PublicSuffixStore& singleton();
+    PublicSuffixStore() = default;
+
+    WEBCORE_EXPORT void enablePublicSuffixCache();
```
Does this ever get called? What is the value of having this?

szewai replied:
oops, forget to add it back to `WebProcess::platformInitializeProcess` while resolving conflicts; will add.

weinig commented on Source/WebCore/platform/PublicSuffix.h:
```diff
@@ -25,13 +25,33 @@
+    PublicSuffixStore() = default;
+
+    WEBCORE_EXPORT void enablePublicSuffixCache();
+    WEBCORE_EXPORT void addPublicSuffix(const String& domain);
```
Does this actually take a domain? Or should it be the public suffix?

Given all the functions taking strings here, consider creating strongly typed structs to encapsulate the concepts. Lots of very similar sounding terms, "domain", "host", "public suffix" for clients to get confused about.

szewai replied:
This should be a public suffix; will update the name of the parameter.
This is good suggestion; I will make a followup to make the functions take or return RegistrableDomain / URL, since it's going to change call sites.

weinig commented on Source/WebCore/platform/PublicSuffix.h:
```diff
@@ -25,13 +25,33 @@
-WEBCORE_EXPORT String topPrivatelyControlledDomain(const String& domain);
-WEBCORE_EXPORT void setTopPrivatelyControlledDomain(const String& domain, const String& topPrivatelyControlledDomain);
-String decodeHostName(const String& domain);
+class PublicSuffixStore {
```
Given the type name, you should probably rename the file to match.

weinig commented on Source/WebCore/platform/PublicSuffix.h:
```diff
@@ -25,13 +25,33 @@
+    WEBCORE_EXPORT void enablePublicSuffixCache();
+    WEBCORE_EXPORT void addPublicSuffix(const String& domain);
+
+    WEBCORE_EXPORT bool isPublicSuffix(StringView domain);
```
If any of these are driven by spec algorithms, I recommend adding comments with links to the specs where you can, to clarify what exactly the functions intention is.

weinig commented on Tools/TestWebKitAPI/Tests/WebCore/PublicSuffix.cpp:
```diff
@@ -75,112 +76,113 @@ TEST_F(PublicSuffix, IsPublicSuffix)
+    EXPECT_EQ(String::fromUTF8("ÅÄÖ"), publicSuffixStore.topPrivatelyControlledDomain(String::fromUTF8("ÅÄÖ")));
+    EXPECT_EQ(String("test.com"_s), publicSuffixStore.topPrivatelyControlledDomain(".test.com"_s));
+    EXPECT_EQ(String(), publicSuffixStore.topPrivatelyControlledDomain("...."_s));
 }
```
You should add tests that utilize the new `enablePublicSuffixCache` and `addPublicSuffix` functionality.

cdumez commented on Source/WebCore/platform/PublicSuffixStore.cpp:
```diff
@@ -0,0 +1,122 @@
+
+PublicSuffixStore& PublicSuffixStore::singleton()
+{
+    static NeverDestroyed<PublicSuffixStore> globalStore;
```
You mentioned in your response to Sam that you want this to work from workers too. However, this singleton implementation is not currently thread-safe. You'd need to use LazyNeverDestroyed + std::call_once.

szewai replied:
Fixed

cdumez commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -4065,8 +4065,9 @@ const URL& Document::urlForBindings() const
-            auto domainString = topPrivatelyControlledDomain(domain.toStringWithoutCopying());
-            auto otherDomainString = topPrivatelyControlledDomain(otherDomain.toStringWithoutCopying());
+            auto& publicSuffixStore = PublicSuffixStore::singleton();
+            auto domainString = publicSuffixStore.topPrivatelyControlledDomain(domain.toStringWithoutCopying());
```
Not new to this patch but I think we should consider updating `topPrivatelyControlledDomain()` to take in a `StringView` instead of a `String`. Having to call `toStringWithoutCopying()` is a usual sign of this.

cdumez commented on Source/WebCore/platform/PublicSuffixStore.h:
```diff
@@ -0,0 +1,59 @@
+class PublicSuffixStore {
+public:
+    WEBCORE_EXPORT static PublicSuffixStore& singleton();
+    PublicSuffixStore() = default;
```
This shouldn't be public. You're going to want to add LazyNeverDestroyed as friend like we usually do.

cdumez commented on Source/WebCore/platform/PublicSuffixStore.h:
```diff
@@ -0,0 +1,59 @@
+    WEBCORE_EXPORT void clearHostTopPrivatelyControlledDomainCache();
+
+    // https://url.spec.whatwg.org/#host-public-suffix
+    WEBCORE_EXPORT bool isPublicSuffix(StringView domain);
```
Ideally this should be `const`, making the cache `mutable` if necessary.

cdumez commented on Source/WebCore/platform/PublicSuffixStore.h:
```diff
@@ -0,0 +1,59 @@
+
+    // https://url.spec.whatwg.org/#host-public-suffix
+    WEBCORE_EXPORT bool isPublicSuffix(StringView domain);
+    WEBCORE_EXPORT String publicSuffix(const String& domain);
```
ditto.

cdumez commented on Source/WebCore/platform/PublicSuffixStore.h:
```diff
@@ -0,0 +1,59 @@
+    // https://url.spec.whatwg.org/#host-public-suffix
+    WEBCORE_EXPORT bool isPublicSuffix(StringView domain);
+    WEBCORE_EXPORT String publicSuffix(const String& domain);
+    WEBCORE_EXPORT String topPrivatelyControlledDomain(const String& host);
```
ditto.

cdumez commented on Source/WebCore/platform/PublicSuffixStore.h:
```diff
@@ -0,0 +1,59 @@
+    WEBCORE_EXPORT String topPrivatelyControlledDomain(const String& host);
+
+private:
+    bool platformIsPublicSuffix(StringView domain);
```
ditto.

cdumez commented on Source/WebCore/platform/PublicSuffixStore.h:
```diff
@@ -0,0 +1,59 @@
+
+private:
+    bool platformIsPublicSuffix(StringView domain);
+    String platformTopPrivatelyControlledDomain(const String& host);
```
ditto.

cdumez commented on Source/WebKit/UIProcess/ProvisionalPageProxy.cpp:
```diff
@@ -286,15 +286,14 @@ void ProvisionalPageProxy::goToBackForwardItem(API::Navigation& navigation, WebB
     m_page->maybeInitializeSandboxExtensionHandle(process(), itemURL, item.resourceDirectoryURL(), sandboxExtensionHandle);
 
-    GoToBackForwardItemParameters parameters { navigation.navigationID(), item.itemID(), *navigation.backForwardFrameLoadType(), shouldTreatAsContinuingLoad, WTFMove(websitePoliciesData), m_page->lastNavigationWasAppInitiated(), existingNetworkResourceLoadIdentifierToResume, topPrivatelyControlledDomain, WTFMove(sandboxExtensionHandle) };
+    auto publicSuffix = WebCore::PublicSuffixStore::singleton().publicSuffix(URL(item.url()).host().toString());
```
We may want to update `publicSuffix()` to take a StringView as well.

cdumez reviewed (approved):
LGTM with changes.

pvollan commented on Source/WebCore/platform/PublicSuffixStore.h:
```diff
@@ -0,0 +1,59 @@
+/*
+ * Copyright (C) 2013 Apple Inc. All rights reserved.
```
2024

pvollan commented on Source/WebCore/platform/PublicSuffixStore.cpp:
```diff
@@ -0,0 +1,126 @@
+    Locker locker { m_publicSuffixCacheLock };
+
+    if (m_publicSuffixCache)
+        m_publicSuffixCache->add(crossThreadCopy(publicSuffix));
```
Should we check if `publicSuffix` is valid with `platformIsPublicSuffix` before adding?

pvollan replied:
Actually, I think the answer is no, since that would make a call to the CFNetwork function we are trying to avoid calling in the WebContent process.

szewai replied:
Yes, the added public suffix should be truth.

pvollan reviewed (approved):
R=me.

annevk commented on Source/WebKit/Shared/GoToBackForwardItemParameters.h:
```diff
@@ -43,7 +43,7 @@ struct GoToBackForwardItemParameters {
     bool lastNavigationWasAppInitiated;
     std::optional<NetworkResourceLoadIdentifier> existingNetworkResourceLoadIdentifierToResume;
-    std::optional<String> topPrivatelyControlledDomain;
+    String publicSuffix;
```
I don't understand why topPrivatelyControlledDomain is being changed to publicSuffix here and in other places. Doesn't that change the scope quite drastically?

szewai replied:
UI process used to send (host , registrable domain) to web process and it is put in the hostTopPrivatelyControlledDomainCache. Now UI process sends public suffix to web process and it's put in publicSuffixCache.
hostTopPrivatelyControlledDomainCache entry helps avoid calling platform function when getting registrable domain for one host; publicSuffixCache entry helps multiple hosts.

szewai committed:
Introduce PublicSuffixStore with cache
https://bugs.webkit.org/show_bug.cgi?id=271188
rdar://124971859

Reviewed by Per Arne Vollan and Chris Dumez.

Intoduce a new class PublicSuffixStore and move existing public suffix related functions into it. PublicSuffixStore can
have a cache for public suffix. When the cache is in use, PublicSuffixStore searches public suffix in it before invoking
platform specific code (this may help us to tighten sandbox in some process later). Currently the cache is only enabled
in web process on Cocoa platforms.

Before this patch, web process has a cache for finding top privately controlled domain of a given host. UI process sends
(host, domain) pair to web process on load time, and web process will add the entry in cache (256605@main). This patch
makes UI process send the public suffix the to web process, and makes web process store it in the new public suffix
cache, since the domain cache has a size limit, and it only benefits search for the given host. Also, this patch fixes
the regression from 270762@main, that web process does not update the cache on receiving LoadParameters (since
setTopPrivatelyControlledDomain is dropped from LoadParameters::decode), by adding addPublicSuffix to
WebPage::platformDidReceiveLoadParameters.

* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/SourcesCocoa.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::urlForBindings const):
* Source/WebCore/loader/ResourceLoadStatistics.cpp:
* Source/WebCore/loader/cache/MemoryCache.cpp:
* Source/WebCore/page/Quirks.cpp:
(WebCore::isYahooMail):
(WebCore::Quirks::shouldHideSearchFieldResultsButton const):
(WebCore::Quirks::isAmazon const):
(WebCore::Quirks::isGoogleMaps const):
* Source/WebCore/page/SecurityOrigin.cpp:
(WebCore::SecurityOrigin::isSameSiteAs const):
(WebCore::SecurityOrigin::isMatchingRegistrableDomainSuffix const):
* Source/WebCore/page/csp/ContentSecurityPolicySourceList.cpp:
(WebCore::ContentSecurityPolicySourceList::isValidSourceForExtensionMode):
* Source/WebCore/platform/Curl.cmake:
* Source/WebCore/platform/PublicSuffixStore.cpp: Added.
(WebCore::PublicSuffixStore::singleton):
(WebCore::PublicSuffixStore::clearHostTopPrivatelyControlledDomainCache):
(WebCore::PublicSuffixStore::isPublicSuffix const):
(WebCore::PublicSuffixStore::publicSuffix const):
(WebCore::PublicSuffixStore::topPrivatelyControlledDomain const):
* Source/WebCore/platform/PublicSuffixStore.h: Renamed from Source/WebCore/platform/PublicSuffix.h.
* Source/WebCore/platform/RegistrableDomain.h:
(WebCore::RegistrableDomain::uncheckedCreateFromHost):
(WebCore::RegistrableDomain::registrableDomainFromHost):
* Source/WebCore/platform/SourcesSoup.txt:
* Source/WebCore/platform/cocoa/PublicSuffixCocoa.mm: Removed.
* Source/WebCore/platform/cocoa/PublicSuffixStoreCocoa.mm: Added.
(WebCore::isPublicSuffixCF):
(WebCore::PublicSuffixStore::platformIsPublicSuffix const):
(WebCore::PublicSuffixStore::platformTopPrivatelyControlledDomain const):
(WebCore::PublicSuffixStore::enablePublicSuffixCache):
(WebCore::PublicSuffixStore::addPublicSuffix):
* Source/WebCore/platform/glib/UserAgentQuirks.cpp:
(WebCore::UserAgentQuirks::quirksForURL):
* Source/WebCore/platform/graphics/skia/PathSkia.cpp:
* Source/WebCore/platform/network/NetworkStorageSession.cpp:
* Source/WebCore/platform/network/ResourceRequestBase.cpp:
(WebCore::ResourceRequestBase::partitionName):
* Source/WebCore/platform/network/cf/NetworkStorageSessionCFNet.cpp:
* Source/WebCore/platform/network/cf/ResourceRequestCFNet.cpp:
* Source/WebCore/platform/network/curl/CookieJarDB.cpp:
(WebCore::CookieJarDB::hasCookies):
(WebCore::CookieJarDB::canAcceptCookie):
* Source/WebCore/platform/network/curl/PublicSuffixStoreCurl.cpp: Renamed from Source/WebCore/platform/network/curl/PublicSuffixCurl.cpp.
(WebCore::PublicSuffixStore::platformIsPublicSuffix const):
(WebCore::topPrivatelyControlledDomainInternal):
(WebCore::PublicSuffixStore::platformTopPrivatelyControlledDomain const):
* Source/WebCore/platform/soup/PublicSuffixStoreSoup.cpp: Renamed from Source/WebCore/platform/soup/PublicSuffixSoup.cpp.
(WebCore::PublicSuffixStore::platformIsPublicSuffix const):
(WebCore::permissiveTopPrivateDomain):
(WebCore::PublicSuffixStore::platformTopPrivatelyControlledDomain const):
* Source/WebKit/NetworkProcess/cocoa/NetworkProcessCocoa.mm:
* Source/WebKit/NetworkProcess/soup/NetworkDataTaskSoup.cpp:
(WebKit::NetworkDataTaskSoup::shouldAllowHSTSPolicySetting const):
* Source/WebKit/Shared/GoToBackForwardItemParameters.h:
* Source/WebKit/Shared/GoToBackForwardItemParameters.serialization.in:
* Source/WebKit/Shared/LoadParameters.h:
* Source/WebKit/Shared/LoadParameters.serialization.in:
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionMatchPatternCocoa.mm:
(WebKit::WebExtensionMatchPattern::hostIsPublicSuffix const):
* Source/WebKit/UIProcess/ProvisionalPageProxy.cpp:
(WebKit::ProvisionalPageProxy::goToBackForwardItem):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::launchProcessForReload):
(WebKit::WebPageProxy::loadRequestWithNavigationShared):
(WebKit::WebPageProxy::loadFile):
(WebKit::WebPageProxy::goToBackForwardItem):
(WebKit::WebPageProxy::resetStateAfterProcessTermination):
* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::WebProcessProxy::processDidTerminateOrFailedToLaunch):
(WebKit::WebProcessProxy::didExceedMemoryFootprintThreshold):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataRecord.cpp:
(WebKit::WebsiteDataRecord::displayNameForHostName):
(WebKit::WebsiteDataRecord::displayNameForOrigin):
(WebKit::WebsiteDataRecord::topPrivatelyControlledDomain):
* Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm:
(WebKit::WebPage::platformDidReceiveLoadParameters):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::goToBackForwardItem):
* Source/WebKit/WebProcess/cocoa/WebProcessCocoa.mm:
(WebKit::prewarmLogs):
(WebKit::WebProcess::platformInitializeProcess):
* Tools/TestWebKitAPI/Tests/WebCore/PublicSuffix.cpp:
(TestWebKitAPI::TEST_F):

Canonical link: https://commits.webkit.org/276505@main
